File: chapitre6/exercice6.2.py
import os
import sys

def main():
    # Vérification du nombre d'arguments [1]
    # sys.argv est le nom du script, donc len(sys.argv) doit être >= 2
    if len(sys.argv) < 2:
        print(f"Usage : {sys.argv} répertoire1 répertoire2 ...", file=sys.stderr)
        sys.exit(1)

    # On boucle sur chaque répertoire passé en argument [1, 2]
    for i in range(1, len(sys.argv)):
        repertoire = sys.argv[i]
        
        try:
            # os.listdir remplace opendir/readdir et retourne la liste des noms [2, 3]
            entrees = os.listdir(repertoire)
        except OSError:
            print(f"Erreur d'ouverture du répertoire : {repertoire}")
            continue

        print(f"\n--- Analyse du répertoire : {repertoire} ---")
        
        for nom_entree in entrees:
            # Construction du chemin complet (remplace sprintf) [4]
            chemin_complet = os.path.join(repertoire, nom_entree)

            try:
                # os.stat remplace la fonction stat() du C [5, 6]
                st = os.stat(chemin_complet)

                # Vérification : fichier ordinaire (S_ISREG) et taille > 1 Mo [7-9]
                # 1 Mo = 1024 * 1024 = 1 048 576 octets
                if os.path.isfile(chemin_complet) and st.st_size > 1048576:
                    # st.st_uid correspond à l'UID du propriétaire [6, 10, 11]
                    print(f"Nom : {nom_entree} | Taille : {st.st_size} octets | UID : {st.st_uid}")
            except OSError:
                # Si le fichier est inaccessible pendant l'analyse
                continue

if __name__ == "__main__":
    main()

File: chapitre6/exercice6.2.c
/*Ecrire un programme qui prend en argument des noms de répertoire et aﬃche
la liste des ﬁchiers de ces répertoires qui ont une taille supérieure à (à peu près) 1M o avec
l’UID du propriétaire du ﬁchier.*/
#include <stdio.h>
#include <dirent.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char** argv) {
    DIR *dir;
    struct dirent *ent;
    struct stat st;
    char chemin;

    if (argc < 2) {
        fprintf(stderr, "Usage : %s répertoire1 répertoire2 ...\n", argv);
        exit(1);
    }

    // On boucle sur chaque répertoire passé en argument
    for (int i = 1; i < argc; i++) {
        dir = opendir(argv[i]); 
        if (dir == NULL) {
            perror("Erreur d'ouverture");
            continue;
        }

        printf("\n--- Analyse du répertoire : %s ---\n", argv[i]);
        while ((ent = readdir(dir)) != NULL) {
            sprintf(chemin, "%s/%s", argv[i], ent->d_name);


            if (stat(chemin, &st) == 0) {
                // On vérifie si c'est un fichier ordinaire et si taille > 1 Mo
                // 1 Mo = 1024 * 1024 = 1 048 576 octets
                if (S_ISREG(st.st_mode) && st.st_size > 1048576) {
                    printf("Nom : %s | Taille : %ld octets | UID : %d\n", 
                           ent->d_name, (long)st.st_size, st.st_uid); // [7, 8]
                }
            }
        }
        closedir(dir);
    }
    return 0;
}